lighthouse-core/gather/gatherers/trace-elements.js

```diff
@@ -49,12 +54,70 @@ function collectTraceElements(attributeMarker) {
 
 class TraceElements extends Gatherer {
   /**
-   * @param {LH.TraceEvent | undefined} lcpEvent
+   * @param {LH.TraceEvent | undefined} event
    * @return {number | undefined}
    */
-  static getNodeIDFromTraceEvent(lcpEvent) {
-    return lcpEvent && lcpEvent.args &&
-      lcpEvent.args.data && lcpEvent.args.data.nodeId;
+  static getNodeIDFromTraceEvent(event) {
+    return event && event.args &&
+      event.args.data && event.args.data.nodeId;
+  }
+
+  /**
+   * @param {Array<number>} rect
+   * @return {LH.Artifacts.Rect}
+   */
+  static traceRectToLHRect(rect) {
+    const rectArgs = {
+      x: rect[0],
+      y: rect[1],
+      width: rect[2],
+      height: rect[3],
+    };
+    return addRectTopAndBottom(rectArgs);
+  }
+
+  /**
+   * @param {Array<LH.TraceEvent>} mainThreadEvents
+   * @return {Array<number>}
+   */
+  static getCLSNodeIdsFromMainThreadEvents(mainThreadEvents) {
+    const clsPerNodeMap = new Map();
+    /** @type {Set<number>} */
+    const clsNodeIds = new Set();
+    const shiftEvents = mainThreadEvents
+      .filter(e => e.name === 'LayoutShift')
+      .map(e => e.args && e.args.data);
+
+    shiftEvents.forEach(event => {
+      if (!event) {
+        return;
+      }
+
+      event.impacted_nodes && event.impacted_nodes.forEach(node => {
+        if (!node.node_id || !node.old_rect || !node.new_rect) {
+          return;
+        }
+
+        const oldRect = TraceElements.traceRectToLHRect(node.old_rect);
+        const newRect = TraceElements.traceRectToLHRect(node.new_rect);
+        const areaOfImpact = getRectArea(oldRect) +
+          getRectArea(newRect) -
+          getRectOverlapArea(oldRect, newRect);
+
+        let prevShiftTotal = 0;
+        if (clsPerNodeMap.has(node.node_id)) {
+          prevShiftTotal += clsPerNodeMap.get(node.node_id);
+        }
+        clsPerNodeMap.set(node.node_id, prevShiftTotal + areaOfImpact);
```

Comment:
i think we should show this value in the report, so people can better understand how significant each of the 5 are. a "CLS contribution" column.

Could that work?

Reply:
It's definitely doable. I think we'd also need to describe somewhere (in the description or maybe a tooltip?) that CLS contribution is the total number of pixels moved or something like that. Thoughts?

Reply:
in that thread we had with skobes, my understanding is these values were in the same unit as CLS is. "shifterinos" basically. 

I'd be fine with presenting it like that or dividing by the CLS and presenting as "% contribution to CLS"

Reply:
I guess the trouble here is the current math used for ranking impact doesn't calculate the individual "shifterinos" per element, but instead just the total impacted area (# of pixels). The math for turning that into "shifterinos" is non-trivial if I recall. Each shift event does have a CLS of all the rects affected by that shift, but there's no way of breaking that down per element either.

Reply:
Ah that could be.

Okay I think we're gonna do a followup here as well to add this second column. I know each layoutshift trace event has the CLS thus far for the page. pretty sure the other number is the total shifterinos for this particular layout.  If that's the case, then we can calculate % of CLS for each element.   I'm pretty sure.

The only thing I don't know about is viewport size, etc.

Reply:
If we are punting this @Beytoven can you make a new issue?